### Q1 Pandas version

In [1]:
import pandas as pd

pd.__version__

'2.3.2'

**Answer**: Pandas version in use is 2.3.2.

### Q2 Records count

In [2]:
df = pd.read_csv("../data/car_fuel_efficiency.csv")
# First let's have a peek what we have in the file.
df.head()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

In [3]:
records = len(df)  # Prints number of records.
print(records)

9704


**Answer**: There is 9704 records in the csv file.

### Q3 Fuel types

In [4]:
fuel = df[
    "fuel_type"
].nunique()  # This print number of unique values in fuel_type column.
print(fuel)

2


**Answer**: There is only two distinct values in the fuel types.

### Q4 Missing values

In [5]:
missing = df.isnull().sum()
print(missing)

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64


**Answer**: Total amount of missing values is 2622, located in four distinct columns.

### Q5 Max fuel efficiency

In [6]:
# Fuel efficiency, cars from asia
cars_asia = df.loc[df["origin"] == "Asia", "fuel_efficiency_mpg"]

# Find the maximum value from the filtered data
max_asia = cars_asia.max()
print(max_asia)

23.759122836520497


**Answer**: Max fuel efficiency in a car made in Asia was approx 23,8.

### Q6 Median value of horsepower

In [7]:
# 1 median of horsepower
horsepow = df["horsepower"].median()
print(horsepow)

149.0


In [8]:
# 2 Find most frequent value in horsepower
frequent = df["horsepower"].mode()[0]
print(frequent)

152.0


In [9]:
# 3 Fill missing values with most frequent one
df["horsepower"] = df["horsepower"].fillna(frequent)

In [10]:
# 4 median of horsepower after filling
horsepow_filled = df["horsepower"].median()
print(horsepow_filled)

152.0


**Answer**: Has the value changed? No, median didn't change in this case.

### Q7 Sum of weights

In [11]:
# 1 Select all the cars from Asia
car_asia_all = df.loc[df["origin"] == "Asia"]
print(car_asia_all)

      engine_displacement  num_cylinders  horsepower  vehicle_weight  \
8                     250            1.0       174.0     2714.219310   
12                    320            5.0       145.0     2783.868974   
14                    200            6.0       160.0     3582.687368   
20                    150            3.0       197.0     2231.808142   
21                    160            4.0       133.0     2659.431451   
...                   ...            ...         ...             ...   
9688                  260            4.0       152.0     3948.404625   
9692                  180            3.0       188.0     3680.341381   
9693                  280            2.0       148.0     2545.070139   
9698                  180            1.0       131.0     3107.427820   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year origin fuel_type         drivetrain  num_doors  \
8             10.3        2016   Asia    Diesel  Front

In [12]:
# 2 Select only columns vehicle_weight and model_year
select_two = car_asia_all[["vehicle_weight", "model_year"]]
print(select_two)

      vehicle_weight  model_year
8        2714.219310        2016
12       2783.868974        2010
14       3582.687368        2007
20       2231.808142        2011
21       2659.431451        2016
...              ...         ...
9688     3948.404625        2018
9692     3680.341381        2016
9693     2545.070139        2012
9698     3107.427820        2005
9703     2908.043477        2005

[3247 rows x 2 columns]


In [13]:
# 3 Select the first 7 values
select_two_seven = select_two.head(7)
print(select_two_seven)

    vehicle_weight  model_year
8      2714.219310        2016
12     2783.868974        2010
14     3582.687368        2007
20     2231.808142        2011
21     2659.431451        2016
34     2844.227534        2014
38     3761.994038        2019


In [14]:
# 4 Get the underlying NumPy array. Let's call it X.
X = select_two_seven.values
print(X)

[[2714.21930965 2016.        ]
 [2783.86897424 2010.        ]
 [3582.68736772 2007.        ]
 [2231.8081416  2011.        ]
 [2659.43145076 2016.        ]
 [2844.22753389 2014.        ]
 [3761.99403819 2019.        ]]


In [15]:
import numpy as np

np.__version__

'2.3.3'

In [16]:
# 5 Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
# Get the transpose of X
X_T = X.T

# Compute the matrix-matrix multiplication XTX
XTX = np.matmul(X_T, X)

print(f"Transpose of X (X.T):\n{X_T}")
print(f"Matrix-matrix multiplication (XTX):\n{XTX}")

Transpose of X (X.T):
[[2714.21930965 2783.86897424 3582.68736772 2231.8081416  2659.43145076
  2844.22753389 3761.99403819]
 [2016.         2010.         2007.         2011.         2016.
  2014.         2019.        ]]
Matrix-matrix multiplication (XTX):
[[62248334.33150762 41431216.5073268 ]
 [41431216.5073268  28373339.        ]]


In [17]:
# 6 Invert XTX.
XTX_inv = np.linalg.inv(XTX)
print(XTX_inv)

[[ 5.71497081e-07 -8.34509443e-07]
 [-8.34509443e-07  1.25380877e-06]]


In [18]:
# 7 Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

In [19]:
# 8 Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
# Multiply the inverse of XTX with the transpose of X
temp = np.matmul(XTX_inv, X_T)
print(temp)

# Multiply the result by y to get the weights
w = np.matmul(temp, y)
print(w)

[[-1.31202622e-04 -8.63909858e-05  3.72634923e-04 -4.02726650e-04
  -1.62513724e-04 -5.52342829e-05  4.65094049e-04]
 [ 2.62636846e-04  1.96990690e-04 -4.73392228e-04  6.58944477e-04
   3.08357831e-04  1.51636137e-04 -6.07979633e-04]]
[0.01386421 0.5049067 ]


In [20]:
# 9 What's the sum of all the elements of the result?
sum_w = np.sum(w)
print(sum_w)

0.5187709081074016


**Answer**: Sum is 0,52.